api/handler/multipart_upload.go

```diff
@@ -0,0 +1,626 @@
+			return
+		}
+		hasData = true
+	}
+
+	if len(r.Header.Get(api.AmzTagging)) > 0 {
+		data.TagSet, err = parseTaggingHeader(r.Header)
+		if err != nil {
+			h.logAndSendError(w, "could not parse tagging", reqInfo, err, additional...)
+			if err = h.obj.AbortMultipartUpload(r.Context(), uploadInfo); err != nil {
+				h.log.Error("could not delete multipart upload init file", additional...)
+			}
+			return
+		}
+		hasData = true
+	}
+
+	metadata := parseMetadata(r)
+	if contentType := r.Header.Get(api.ContentType); len(contentType) > 0 {
+		metadata[api.ContentType] = contentType
+	}
+
+	p := &layer.UploadPartParams{
+		Info:       uploadInfo,
+		PartNumber: 0,
+		Header:     metadata,
+	}
+
+	if hasData {
+		b, err = json.Marshal(data)
+		if err != nil {
+			h.logAndSendError(w, "could not marshal json with acl and/or tagging", reqInfo, err, additional...)
+			return
+		}
+		p.Reader = bytes.NewReader(b)
+	}
+
+	info, err := h.obj.UploadPart(r.Context(), p)
+	if err != nil {
+		h.logAndSendError(w, "could not upload a part", reqInfo, err, additional...)
+	}
+
+	resp := InitiateMultipartUploadResponse{
+		Bucket:   info.Bucket,
+		Key:      info.Headers[layer.UploadKeyAttributeName],
+		UploadID: info.Headers[layer.UploadIDAttributeName],
+	}
+
+	if err := api.EncodeToResponse(w, resp); err != nil {
+		h.logAndSendError(w, "could not encode InitiateMultipartUploadResponse to response", reqInfo, err, additional...)
+		return
+	}
+}
+
+func (h *handler) UploadPartHandler(w http.ResponseWriter, r *http.Request) {
+	reqInfo := api.GetReqInfo(r.Context())
+
+	bktInfo, err := h.obj.GetBucketInfo(r.Context(), reqInfo.BucketName)
+	if err != nil {
+		h.logAndSendError(w, "could not get bucket info", reqInfo, err)
+		return
+	}
+
+	if err := checkOwner(bktInfo, r.Header.Get(api.AmzExpectedBucketOwner)); err != nil {
+		h.logAndSendError(w, "expected owner doesn't match", reqInfo, err)
+		return
+	}
+	var (
+		queryValues = r.URL.Query()
+		uploadID    = queryValues.Get(uploadIDHeaderName)
+		additional  = []zap.Field{zap.String("uploadID", uploadID), zap.String("Key", reqInfo.ObjectName)}
+	)
+
+	partNumber, err := strconv.Atoi(queryValues.Get(partNumberHeaderName))
+	if err != nil || partNumber < layer.UploadMinPartNumber || partNumber > layer.UploadMaxPartNumber {
+		h.logAndSendError(w, "invalid part number", reqInfo, errors.GetAPIError(errors.ErrInvalidPartNumber))
+		return
+	}
+
+	p := &layer.UploadPartParams{
+		Info: &layer.UploadInfoParams{
+			UploadID: uploadID,
+			Bkt:      bktInfo,
+			Key:      reqInfo.ObjectName,
+		},
+		PartNumber: partNumber,
+		Size:       r.ContentLength,
+		Reader:     r.Body,
+	}
+
+	info, err := h.obj.UploadPart(r.Context(), p)
+	if err != nil {
+		h.logAndSendError(w, "could not upload a part", reqInfo, err, additional...)
+		return
+	}
+
+	w.Header().Set(api.ETag, info.HashSum)
+	api.WriteSuccessResponseHeadersOnly(w)
+}
+
+func (h *handler) UploadPartCopy(w http.ResponseWriter, r *http.Request) {
```

Comment:
It seems you've missed handling the following headers:
* `x-amz-source-expected-bucket-owner`
* `x-amz-copy-source-if-match`
* `x-amz-copy-source-if-modified-since`
* `x-amz-copy-source-if-none-match`
* `x-amz-copy-source-if-unmodified-since`

They should be handle the similar way as `CopyObjectHandler` does.

Reply:
Oh, planned to add this, but forgot. Thanks, fixed.

Reply:
It seems `x-amz-source-expected-bucket-owner` header still isn't handled